src/store/fs/util/entity_manager.rs

```diff
@@ -0,0 +1,1322 @@
+                        (spawn.f)(SpawnArg::Dead).await;
+                    }
+                }
+            };
+            task
+        }
+
+        /// This function needs to be polled by the owner of the actor state to advance the
+        /// entity manager state machine. If it returns a future, that future must be spawned
+        /// by the caller.
+        #[must_use = "this function may return a future that must be spawned by the caller"]
+        pub async fn tick(&mut self) -> Option<impl Future<Output = ()> + Send + 'static> {
+            if let Some(cmd) = self.internal_recv.recv().await {
+                match cmd {
+                    InternalCommand::Shutdown(Shutdown { id, receiver }) => {
+                        let Some(entity_handle) = self.live.remove(&id) else {
+                            error!("Received shutdown command for unknown entity actor {id:?}");
+                            return None;
+                        };
+                        let EntityHandle::Live { send } = entity_handle else {
+                            error!(
+                                "Received shutdown command for entity actor {id:?} that is already shutting down"
+                            );
+                            return None;
+                        };
+                        self.live.insert(
+                            id.clone(),
+                            EntityHandle::ShuttingDown {
+                                send,
+                                recv: receiver,
+                            },
+                        );
+                    }
+                    InternalCommand::ShutdownComplete(ShutdownComplete { state, tasks }) => {
+                        let id = state.id.clone();
+                        let Some(entity_handle) = self.live.remove(&id) else {
+                            error!(
+                                "Received shutdown complete command for unknown entity actor {id:?}"
+                            );
+                            return None;
+                        };
+                        let EntityHandle::ShuttingDown { send, recv } = entity_handle else {
+                            error!(
+                                "Received shutdown complete command for entity actor {id:?} that is not shutting down"
+                            );
+                            return None;
+                        };
+                        // re-assemble the actor from the parts
+                        let mut actor = entity_actor::Actor {
+                            main: self.internal_send.clone(),
+                            recv,
+                            state,
+                            tasks,
+                        };
+                        if actor.recv.is_empty() {
+                            // No commands during shutdown, we can recycle the actor.
+                            self.recycle(send, actor);
```

Comment:
Is there a race condition here?
I struggle to understand the whole web of senders, receivers and spawns and whether there's potential for true parallelism or just concurrency... But in theory `recv` might receive an item after `is_empty()` returns true and before `self.recycle`, right?

What's the worst that could happen in that case?

Otherwise if it's impossible, perhaps it's also impossible for `actor.recv.is_empty()` to return false, in that case maybe there's potential for removing the `recv`. Not sure though.